cdap-api/src/main/java/co/cask/cdap/internal/workflow/ProgramWorkflowAction.java

```diff
@@ -42,11 +50,16 @@
   private String programName;
   private Runnable programRunner;
   private SchedulableProgramType programType;
+  private Map<String, Dataset> datasets;
+  private TransactionSystemClient txClient;
 
-  public ProgramWorkflowAction(String name, String programName, SchedulableProgramType programType) {
+  public ProgramWorkflowAction(String name, String programName, SchedulableProgramType programType,
+                               @Nullable TransactionSystemClient txClient) {
```

Comment:
We decided to make a custom action another "program", because otherwise we do not have access to the datasets in WorkflowDriver's executeAction.  The datasets are only instantianted in the initialize method when we create the WorkflowActionContext's, so only then can we pass the datasets into the ProgramWorkflowAction where they are added to the transaction.